### ISSUE TITLE
Counter re-initialization inside for loop confuses pyright

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2664]
**Describe the bug**
In the following [advent of code](https://adventofcode.com/2021/day/6) solution, I get errors like the following on all lines in the Counter:
`error: Argument of type "Literal[1]" cannot be assigned to parameter "__k" of type "_KT@dict" in function "__getitem__"`

```python
from collections import Counter
fish = Counter(map(int, next(open("input.txt")).split(",")))
for i in range(256):
    fish = Counter(
        {
            0: fish[1],
            1: fish[2],
            2: fish[3],
            3: fish[4],
            4: fish[5],
            5: fish[6],
            6: fish[7] + fish[0],
            7: fish[8],
            8: fish[0],
        }
    )
print(sum(fish.values()))
```
**To Reproduce**
Run pyright on the above code.

**Expected behavior**
No error without strict mode even if pyright can't infer the type. I can get rid of the error by hinting Counter[int], but this error is bubbling up to people in IDEs since it's outside of strict mode.

**VS Code extension or command-line**
pyright 1.1.193 Linux commandline

**Additional context**
It doesn't seem to have a problem if I move the code outside the for loop. I think I've run into this problem before where re-initializing a dict inside a for loop seems to confuse pyright a bit, but I'm not sure if it's related.

**Comment by erictraut** [comment]
Could you please provide a complete, minimal repro? I don't know where the `Counter` symbol comes from or how it's declared. Thanks!

**Comment by GBeauregard** [comment]
> Could you please provide a complete, minimal repro? I don't know where the `Counter` symbol comes from or how it's declared. Thanks!

oh, sorry! I forgot to include `from collections import Counter`

**Comment by erictraut** [comment]
Thanks, I'm able to repro the problem and will investigate further.

**Comment by matejcik** [comment]
I found a possible duplicate, so I'll just leave my reproducer here. I'll open a separate issue if it turns out to be a different thing.

In my case, assigning to a variable inside a loop _widens_ the type of the variable, which breaks subsequent type checking
```python
class State:
    def confirm_dialog(self) -> "State" | bool:
        return False

state = State()
reveal_type(state)  # info: Type of "state" is "State"

for _ in range(1):
    result = state.confirm_dialog()  # error: Cannot access member "confirm_dialog" for type "bool"
    if isinstance(result, State):
        reveal_type(state)  # info: Type of "state" is "State | bool"
        reveal_type(result)  # info: Type of "result" is "State | Unknown"
        state = result
    else:
        reveal_type(state)  # Type of "state" is "State | bool"
        reveal_type(result)  # Type of "result" is "bool | Unknown"
```

**Comment by erictraut** [comment]
@matejcik, this is an an unrelated issue, so I'd appreciate it if you could open a separate issue.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.194, which I just published. It will also be included in the next release of pylance.
